smart-contracts/wasm-transform/src/artifact.rs

```diff
@@ -804,378 +1244,503 @@ impl Handler<&OpCode> for BackPatch {
+                                              // are just playing it safe.
+                if let Some(reserve) = reserve {
+                    self.out.push(Copy);
+                    self.out.push_i32(idx); // from
+                    self.push_loc(reserve); // to
+                }
+                if matches!(opcode, OpCode::LocalSet(..)) {
+                    match last_provide {
+                        // if we had to copy to a reserve location then it is not
+                        // possible to short circuit the instruction.
+                        // We need to insert an additional copy instruction.
+                        Some(back_loc) if reserve.is_none() => {
+                            // instead of inserting LocalSet, just tell the previous
+                            // instruction to copy the value directly into the local.
+                            self.out.back_patch(back_loc, idx as u32)?;
+
+                            // And clear the provider from the stack
+                            self.providers_stack.consume()?;
+                        }
+                        _ => {
+                            self.out.push(Copy);
+                            let _operand = self.push_consume()?; // value first.
+                            self.out.push_i32(idx); //target second
+                        }
+                    }
+                } else {
+                    match last_provide {
+                        // if we had to copy to a reserve location then it is not
+                        // possible to short circuit the instruction since there is an extra Copy
+                        // instruction. We need to insert an additional copy
+                        // instruction.
+                        Some(back_loc) if reserve.is_none() => {
+                            // instead of inserting LocalSet, just tell the previous
+                            // instruction to copy the value directly into the local.
+                            self.out.back_patch(back_loc, idx as u32)?;
+
+                            // And clear the provider from the stack
+                            self.providers_stack.consume()?;
+                            self.providers_stack.provide_existing(Provider::Local(idx));
+                        }
+                        _ => {
+                            // And clear the provider from the stack
+                            self.out.push(Copy);
+                            let _source = self.push_consume()?;
+                            self.out.push_i32(idx); //target second
+                            self.providers_stack.provide_existing(Provider::Local(idx));
+                        }
+                    }
+                }
             }
             OpCode::GlobalGet(idx) => {
+                // TODO: In principle globals could also be "providers" or locations to write.
```

Comment:
No. I had this implemented also, but it adds a bunch more complexity and does not improve performance in the examples I've tried. So I did not add it. I'll revise the comment to add this context instead of a TODO.